### DISCUSSION TITLE
Customize opentelemetry-javaagent including library as dependency

### DISCUSSION THREAD

**Discussion by rohitjain010292** [open-telemetry/opentelemetry-java-instrumentation#6825] [Q&A] [answered]
Hello,

I need to customize default opentelemetry-javaagent to include extra wrapper features. 

As per my understanding there are 2 ways of customizing opentelemetry-javaagent. 

1. Fork repository [opentelemetry-java-instrumentation](https://github.com/open-telemetry/opentelemetry-java-instrumentation) and add our custom code on top of it.
2. Add agent jar (https://mvnrepository.com/artifact/io.opentelemetry.javaagent/opentelemetry-javaagent/1.18.0) as a dependency, then add our custom code on top of it and then build agent jar bundling our custom code and opentelemetry-javaagent jar.
   In this approach, opentelemetry-javaagent library dependency will be:
 ```
 <!-- https://mvnrepository.com/artifact/io.opentelemetry.javaagent/opentelemetry-javaagent -->
<dependency>
    <groupId>io.opentelemetry.javaagent</groupId>
    <artifactId>opentelemetry-javaagent</artifactId>
    <version>1.18.0</version>
    <scope>runtime</scope>
</dependency>
```

In my observation, most of the customizations are adopting approach 1. Approach 1 needs dealing with code rebasing and dealing with code conflicts to get updates in opentelemetry-java-instrumentation repo. Whereas with Approach 2, upgrading is just updating the opentelemetry-javaagent library version number.

What is the preferred approach for customizing opentelemetry-javaagent?

Also for Approach 2, will `io.opentelemetry.javaagent.opentelemetry-javaagent` jar dependency be sufficient? This jar includes instrumentation rules classes as shaded classes and uses a custom classloader to load shaded classes. If we added custom instrumentation rules, will they work along with already shaded classes in the library? I don't have good understanding of how classloaders work. So, I am not sure if this works.

Please provide your suggestions. Thank you.

**Comment by laurit** [comment, marked as answer]
2. is the preferred way, see https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/examples/distro
